### ISSUE TITLE
Circular Depdency on Typescript 4.0 seems invalid

### ISSUE THREAD

**Issue by khrome83** [denoland/deno#7280]
After upgrading to Deno 1.3.2 with TS 4.0, I ran into a strange issue.

5 of my tests started to fail. 

Each test runs a piece of code that uses a dynamic import. The URL for this import is not tagged, as the code is not released. 
That code then has a reference to types in the original project (the one running the tests), but once again off the master branch. Which was Deno version 1.3.0.

This causes the download of those dynamic imports to fail within TypeScript 4.0. Since it's referencing the branch I am trying to upgrade, I can't upgrade the branch in the correct way. Meaning I am unable to open an MR that will pass tests, as tests fail due to the circular reference. 

This would happen if my branches were tagged also, as an older version not built for TS4.0 would throw errors, and the MR that made reference would need to be upgraded. But would fail tests without being able to upgrade through proper channels. 

This seems to force a mono repo pattern so dependencies can be out of sync. But I feel it points to a larger issue as any 3rd party dependency can cause this.

I am not sure of the solution here. Should there by a TS engine reference where each repository can run types against its specific version? It just seems strange to fail due to type issues and not legit code or breaking feature issues.

**Comment by kitsonk** [comment]
> This causes the download of those dynamic imports to fail within TypeScript 4.0.

This whole things needs a bit more detail.  It feels the use case is valid, but really need to get it to be something reproducible.  I am not aware of anything in TypeScript 4.0 that would specifically cause this regression either.

**Comment by khrome83** [comment]
Apologies, this is confusing. I spent half a day figuring out what was happening.

`deno info` does not pull dynamic inputs so it took a long time to figure out where this was happening at. 

Repo 1 - master branch - references a 3rd party dependency that uses a breaking change in TS 4.0 (delete non-optional)
Repo 1 - upgrade branch - fixes all local delete non-optional issues, and upgrades the 3rd party dependency. 
Repo 1 - upgrade branch - references a plugin in another repo (Repo 2) via a dynamic import
Repo 2 - master branch - reference Repo 1 master branch for types only

When running `deno test` on the upgrade branch for Repo 1, the tests that pull the plugin in for testing, also pull in an older version (master) of the same repo. This causes a circular dependency issue, which has TS throw an error on running on Repo 1. 

In production, these would not be master branches, but specific tags, but the issues still exist. I would have had to either not built my plugins in separate repositories or had known the 3rd party dependency that had the breaking change needed to be upgraded first, and SEPARATE from upgrading deno/ts. 

Because the circular reference is related to types, it feels highly reproducible, though specific to projects that build plugin communities. 

Did this help?

The solution BTW was a force pushing directly to master, knowing tests fail and it would not pass CI.

**Comment by kitsonk** [comment]
>  which has TS throw an error on running on Repo 1

It is really hard to guess at what these errors are.  I am afraid it doesn't really help.  You reproducing the issue doesn't help anyone else reproduce the issue.

Also, are you sure you aren't dealing with issues related to how `deno.land/x/` works? https://deno.land/posts/registry2 There were changes where if you don't request a specific version, it doesn't serve the master branch anymore, but instead serve the latest tag.

**Comment by khrome83** [comment]
@kitsonk like I said above, it is the BREAKING CHANGE in TS 4.0 for deleting a non-optional value off a type. It just throws the TS error.

This has nothing to do with deno.land/x/. 

Basically, if a repository has a DYNAMIC import to another repository, that then links BACK to the original, but a different tag or master, you can see this being caused.

Honestly, you could have a dynamic import of the same repository, but a different tag / master.

The version you are upgrading locally has the TS4.0 issues fixed, but the version that is linked too does not. In my situation, it was a 3rd party module that was the offender, and the version I am linking too used that third party version that failed under TS 4.0.

In other words, we do type checking on dynamic imports, which causes you to be in a cycle where you can't pass tests to upgrade.

**Comment by kitsonk** [comment]
> In other words, we do type checking on dynamic imports,

That is working as designed, imports are type checked even dynamic ones if they can be statically determined.  If you don't want to type check your code, use `--no-check`.

**Comment by khrome83** [comment]
So what would the upgrade path look like here? Since the import circles back to the older version that is not TS 4.0? Is there a recommended pattern to avoid that? I understand you can do --no-check, but that does not really work for CI.
